internal/updatecheck/update.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2020 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package updatecheck
+
+import (
+	"encoding/json"
+	"math/rand"
+	"net/http"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+
+	"sigs.k8s.io/krew/internal/installation/semver"
+	"sigs.k8s.io/krew/internal/version"
+)
+
+const (
+	githubVersionURL = "https://api.github.com/repos/kubernetes-sigs/krew/releases/latest"
+)
+
+// for testing
+var versionURL = githubVersionURL
+
+// LatestTag returns the tag name of the latest release on GitHub or "" at random.
+// For development builds, it always returns "".
+func LatestTag() (string, error) {
+	ourTag := version.GitTag()
+	if !isSemver(ourTag) || // no upgrade check for dev builds
+		rand.Float64() > 0.4 { // only do the upgrade check randomly
+		return "", nil
+	}
```

Comment:
please don't do these checks here. try to do them before getting into logic.

```
// LatestTag returns the tag name of the latest release on GitHub or "" at random.
// For development builds, it always returns "".
```

this doesn't reflect what this method does accurately because of this.

Methods ideally should do "fewer things" that are unit testable. This isn't.